static/app/api/apiOptions.spec.tsx

```diff
@@ -0,0 +1,259 @@
+  });
+
+  test('should not modify already encoded parameters', () => {
+    const options = apiOptions('/search/$query/', {
+      staleTime: 0,
+      path: {query: 'test%20query'},
+    });
+
+    expect(options.queryKey[0]).toBe('/search/test%20query/');
+  });
+
+  test('should not include empty options in queryKey', () => {
+    const options = apiOptions('/projects/$id/', {
+      staleTime: 0,
+      path: {id: '123'},
+    });
+
+    expect(options.queryKey).toEqual(['/projects/123/']);
+  });
+
+  test('should stringify number path params', () => {
+    const options = apiOptions('/items/$id/', {
+      staleTime: 0,
+      path: {id: 123},
+    });
+
+    expect(options.queryKey[0]).toBe('/items/123/');
+  });
+
+  test('should not do accidental replacements', () => {
+    const options = apiOptions('/projects/$id1/$id', {
+      staleTime: 0,
+      path: {id: '123', id1: '456'},
+    });
+
+    expect(options.queryKey).toEqual(['/projects/456/123']);
+  });
+
+  test('should extract content data per default', async () => {
+    const options = apiOptions.as<string[]>()('/projects/', {
+      staleTime: 0,
+    });
+
+    MockApiClient.addMockResponse({
+      url: '/projects/',
+      body: ['Project 1', 'Project 2'],
+    });
+
+    const {result} = renderHook(() => useQuery(options), {wrapper});
+
+    await waitFor(() => result.current.isSuccess);
+
+    expect(result.current.data).toEqual(['Project 1', 'Project 2']);
+  });
+
+  test('should extract headers', async () => {
+    const options = apiOptions.as<string[]>()('/projects/', {
+      staleTime: 0,
+    });
+
+    MockApiClient.addMockResponse({
+      url: '/projects/',
+      body: ['Project 1', 'Project 2'],
+      headers: {
+        Link: 'my-link',
+        'X-Hits': 'some-hits',
+      },
+    });
+
+    const {result} = renderHook(
+      () =>
+        useQuery({...options, select: selectWithHeaders(['Link', 'X-Hits'] as const)}),
+      {wrapper}
+    );
+
+    await waitFor(() => result.current.isSuccess);
+
+    expect(result.current.data).toEqual({
+      content: ['Project 1', 'Project 2'],
+      headers: {Link: 'my-link', 'X-Hits': 'some-hits'},
+    });
+
+    // headers should be narrowly typed
+    expectTypeOf(result.current.data!.headers).toEqualTypeOf<{
+      Link: string | undefined;
+      'X-Hits': string | undefined;
+    }>();
+  });
+
+  describe('types', () => {
+    test('should infer types of known API paths', () => {
+      const options = apiOptions(
+        '/projects/$orgSlug/$projectSlug/releases/$releaseVersion/',
+        {
+          staleTime: 0,
+          path: {
+            orgSlug: 'my-org',
+            projectSlug: 'my-project',
+            releaseVersion: 'v1.0.0',
+          },
+        }
+      );
+
+      // todo: Replace `never` with the actual type of apiDefinition
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should always require staleTime', () => {
```

Comment:
one thing I’m not entirely sold on is the fact that `apiOptions` (like `useApiQuery`) always requires us to pass a `staleTime`, because it means when I want to write `apiOptions` for `queryClient.getQueryData`, I also have to provide a `staleTime` (which wouldn’t technically be needed).

If we think that we’ll always further abstract them into something like:

```
const releaseApiOptions = () => apiOptions(...)
```

then it’s probably fine, but not sure if that will always be the case...